sips/sip-merge-staked-sui.md

```diff
@@ -0,0 +1,133 @@
+
+    // === Public getters ===
+    public fun fungible_stake_value(fungible_stake: &FungibleStake): u64 {
+        fungible_stake.value
+    }
+
+    public fun fungible_stake_pool_id(fungible_stake: &FungibleStake): ID {
+        fungible_stake.pool_id
+    }
+
+    public fun fungible_stake_to_sui_amount(pool: &StakingPool, fungible_stake_amount: u64): u64;
+    public fun sui_to_fungible_stake_amount(pool: &StakingPool, sui_amount: u64): u64;
+
+    public fun join_fungible_stake(self: &mut FungibleStake, other: FungibleStake);
+    public fun split_fungible_stake(self: &mut FungibleStake, amount: u64): FungibleStake;
+
+    /// Burn an FungibleStake object to obtain the underlying SUI.
+    public(package) fun redeem_fungible_stake(pool: &mut StakingPool, fungible_stake: FungibleStake, ctx: &TxContext) : Balance<SUI>;
+
+    /// Convert the given staked SUI to an FungibleStake object
+    public(package) fun convert_to_fungible_stake(pool: &mut StakingPool, staked_sui: StakedSui, ctx: &mut TxContext) : FungibleStake;
+
+```
+
+[validator.move](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/sui-system/sources/validator.move)
+
+```move
+    public(package) fun convert_to_fungible_stake(self: &mut Validator, staked_sui: StakedSui, ctx: &TxContext) : FungibleStake;
+    public(package) fun redeem_fungible_stake(self: &mut Validator, fungible_stake: FungibleStake, ctx: &TxContext) : Balance<SUI>;
+```
+
+[validator_set.move](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/sui-system/sources/validator_set.move)
+
+```move
+    public(package) fun convert_to_fungible_stake(self: &mut ValidatorSet, staked_sui: StakedSui, ctx: &TxContext) : FungibleStake;
+    public(package) fun redeem_fungible_stake(self: &mut ValidatorSet, fungible_stake: FungibleStake, ctx: &TxContext) : Balance<SUI>;
+```
+
+[sui_system_state_inner.move](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/sui-system/sources/sui_system_state_inner.move)
+
+```move
+    public(package) fun convert_to_fungible_stake(self: &mut SuiSystemStateInnerV2, staked_sui: StakedSui, ctx: &TxContext) : FungibleStake;
+    public(package) fun redeem_fungible_stake(self: &mut SuiSystemStateInnerV2, fungible_stake: FungibleStake, ctx: &TxContext) : Balance<SUI>;
+```
+
+It's possible I missed some getter functions, but I think this is the gist of it.
+
+## Rationale
+
+### Instead of the FungibleStake object, why don't we just mint a new "LST" Coin per StakingPool?
+
+This is possible and I really like this idea. The main benefit of doing this is that it would immediately enable single-validator LSTs for the entire validator set, which would be pretty cool.
+
+The tricky part would be managing a unique Coin type + metadata per StakingPool. Some notes:
+- We would need a one time witness per StakingPool to create the Coin type. This can be passed into the StakingPool on instantiation.
```

Comment:
Right OTWs are only available in `init` function at package publishing times so I'm not sure if this solution is feasible without making some OTW exceptions.

Reply:
An idea: can we leverage TTO to support this for both newly created and existing `StakingPool` objects? E.g.,

Add something like: 

```
public fun add_lst_coin<T>(cap: Receiving<TreasuryCap<T>>, metadata: Receiving<T>, pool: &mut StakingPool) {
  // assert cap.total_supply == 0 to ensure there has not been any preminting
  // add cap to pool.extra_fields["lst_cap"], which ensures this is the only coin type associated with this pool
  // update CoinMetadata fields to whatever standards we want for LST tokens
  // share metadata
}
```

to `staking_pool`, and then anyone can publish a package that calls `create_currency` in the initializer, then transfers the `TreasuryCap` and `CoinMetadata` to the `StakingPool` object.

This is a bit different than the idea of calling `create_currency` inside staking pool, but it won't require OTW exceptions (which are a bit scary), and I think it satisfies the requirements of making this permissionless + compatible with existing `StakingPool`'s

Reply:
So, after prototyping an LST on top of my reference implementation I'm kind of against making this a Coin, as I would not be able to store Coins from different validators together in the same vector. This is needed to efficiently do unstake operations (LST -> vector<StakedSui>). If we made FungibleStake a Coin, I would have to do a linear amount of dynamic field reads which doesn't feel good